builders/build/substrate-api/py-substrate-interface.md

```diff
@@ -0,0 +1,311 @@
+
+# Construct the API provider
+ws_provider = SubstrateInterface(
+    url="{{ networks.moonbase.wss_url }}",
+) 
+
+def subscription_handler(obj, update_nr, subscription_id):
+
+    print(f"New block #{obj['header']['number']} produced by {obj['author']}")
+
+    if update_nr > 10:
+        return {'message': 'Subscription will cancel when a value is returned', 'updates_processed': update_nr}
+
+
+block_header = ws_provider.subscribe_block_headers(subscription_handler, include_author=True)
+```
+
+### Query for Storage Information {: #query-for-storage-information }
+
+You can use the [`get_metadata_storage_functions`](https://polkascan.github.io/py-substrate-interface/#substrateinterface.SubstrateInterface.get_metadata_storage_functions){target=_blank} to see a list of available storage functions within Moonbeam network's metadata. 
+
+Chain states that are provided in the metadata through storage functions can be queried through the [`query`](https://polkascan.github.io/py-substrate-interface/#substrateinterface.SubstrateInterface.query){target=_blank} method. 
+
+System modules can be queried to provide basic information such as account nonce and balance. The available functions are read from the metadata dynamically, so it's also possible to query for storage information on Moonbeam custom modules such as `ParachainStaking` for state information that's specific to Moonbeam. 
+
+```python
+# Import
+from substrateinterface import SubstrateInterface
+
+# Construct the API provider
+ws_provider = SubstrateInterface(
+    url="{{ networks.moonbase.wss_url }}",
+) 
+
+# List of available storage functions in the metadata
+method_list = ws_provider.get_metadata_storage_functions()
+print(method_list)
+
+# Query basic account information
+account_info = ws_provider.query(
+    module='System',
+    storage_function='Account',
+    params=['0x578002f699722394afc52169069a1FfC98DA36f1']
+)
+# Log the account nonce
+print(account_info.value['nonce'])
+# Log the account free balance
+print(account_info.value['data']['free'])
+
+# Query candidate pool information from Moonbeam's Parachain Staking module
+candidate_pool_info = ws_provider.query(
+    module='ParachainStaking',
+    storage_function='CandidatePool',
+    params=[]
+)
+print(candidate_pool_info)
+```
+
+## Signing and Transactions {: #signing-and-transactions }
+
+### Creating a Keypairs {: #creating-a-keypair }
+
+The Keypair object in Python Substrate Interface is used in the signing of any data, whether it's a transfer, a message, or a contract interaction.  
+
+You can create a Keypair instance from the shortform private key or from the mnemonic. For Moonbeam networks, you also need to specify the KeypairType to be `KeypairType.ECDSA`.
+
+#### Signing Data {: #signing-data }
+
+You can sign any arbitrary data using the keypair through the [`sign`](https://polkascan.github.io/py-substrate-interface/#substrateinterface.Keypair.sign){target=_blank} method.
```

Comment:
same is there a code snippet for this?

also does it need to be a #### can it be a ### or maybe just included int he create a keypair section?

Reply:
Expanded this section to offline signing then commented it out for now. They need to add support for generating the keypair from an Ethereum style address or public key for this to work for us. Will file an issue on their repo, then revisit.